### PR TITLE
ENH: reduce np.zeros, ones, and full array size with broadcasting

### PR DESCRIPTION
**Pull request by Carreau**
As array are not mutable, we can actually back large uniform array by a
single value.

In [3]: %time %memit N= 50000; da.zeros((N, N), dask_broadcast_trick=True).sum().compute()
peak memory: 65.86 MiB, increment: 1.66 MiB
CPU times: user 1.69 s, sys: 42.7 ms, total: 1.73 s
Wall time: 1.21 s

In [4]: %time %memit N= 50000; da.zeros((N, N), dask_broadcast_trick=False).sum().compute()
peak memory: 364.25 MiB, increment: 298.34 MiB
CPU times: user 8.75 s, sys: 6.13 s, total: 14.9 s
Wall time: 6.55 s

That should probably be folded into wrap_func_shape_as_first_arg, and not change the function/task names(?), Let me know what you think and I can also look into testing more, and see if I can extend it to the `x_like` functions.

--- 

- [ ] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by mrocklin** [comment]
I'm fine with this in principle, but I'm not a expert with these tricks.  I suspect that @jakirkham might have a sense for fail cases that might occur.

**Comment by Carreau** [comment]
> I suspect that @jakirkham might have a sense for fail cases that might occur.

Current "failures" case I was thinking of was:

- Assign one value assigns everywhere (that does not seem to happen on Arrays, and chunks should read only if you access them directly), 
- "C" or "F" order is not respected (I haven't tested), not sure it matters with an array whose values are all the same and is immutable, but maybe this info get lost at reification ?  
- somehow the informations that this is a broadcasting trick get lost and it get reified
   - include Rechunking, that will most likely do that.

If you know of anything that get directly given a pointer to the memory of each chunk and really expects a real numpy array, it will likely fail is my guess, but I'm not too sure what could do that.

**Comment by Carreau** [comment]
So I tested locally a bit more, and that does reduce the data transfer between workers and memory consumption; I've cleanup the implementation, extended to also apply to `full`. 

I haven't touched `empty_like` as it's likely to be used directly for mutation internally and I see two implementation in both `wrap.py` and `creation.py` and I'm unsure which one is meant for what.

File: dask/array/wrap.py [hunk 1]
```diff
@@ -134,9 +134,40 @@ def wrap(wrap_func, func, **kwargs):
...
-zeros = w(np.zeros, dtype="f8")
-empty = w(np.empty, dtype="f8")
+
+def broadcast_trick(func):
+    """
+    Provide a decorator to wrap common numpy function with a broadcast trick.
+
+    Dask arrays are currently immutable; thus when we know an array is uniform,
+    we can replace the actual data by a single value and have all elements point
+    to it, this reducing the size.
```
Comment (jsignell) [inline dask/array/wrap.py:144]:
```suggestion
    to it, thus reducing the size.
```

Comment (Carreau) [reply dask/array/wrap.py]: thanks !

**Comment by jsignell** [comment]
I tried this out locally and it seems to work as expected. I imagine you are still adding test @Carreau. Otherwise we can ping someone for a review.

**Comment by Carreau** [comment]
I'm not too sure which test to add. Is there a way to test the number of bytes transferred between workers for a given workflow is less than a threshold for a given computation ?

**Comment by jakirkham** [comment]
Thanks for working on this Matthias! 😄

In the context of broadcasting array element, I wonder if `empty` shouldn't just be an alias for `zeros`. Broadcasting an uninitialized value through an array is probably similar in performance to doing the same thing with a `0`.

**Comment by jakirkham** [comment]
cc @pentschev (if you have any thoughts here 🙂)

**Comment by Carreau** [comment]
I don't know, I actually don't see any reason for empty to be used beyond mutation to initialize something else; So usage of empty are likely mistakes ? 

But maybe that's separate discussion, and maybe reading from empty in user-code should emit a warning ?

**Comment by Carreau** [comment]
> In the context of broadcasting array element, I wonder if `empty` shouldn't just be an alias for `zeros`

Or `full(NaN)`? (when possible)  This is the "safest" value as it would poison anything that touches it and seem like the easiest to detect bugs.

**Comment by pentschev** [comment]
Changes here make sense to me, I think from the `__array_function__` perspective there's nothing we would need to do. Currently we can't dispatch on `empty` and derived functions, but with [NEP-35](https://numpy.org/neps/nep-0035-array-creation-dispatch-with-array-function.html) we may be able to add the `like=` argument there and it seems like there would be no problem for that to work as well.

> Broadcasting an uninitialized value through an array is probably similar in performance to doing the same thing with a 0.
> ...
> ..., and maybe reading from empty in user-code should emit a warning ?

Making `empty` an alias for `zeros` would still potentially lead to incorrect results. I think the warning could be more useful in that sense, although that isn't part of NumPy and one could very much argue this should be known by users.

**Comment by Carreau** [comment]
> Making `empty` an alias for `zeros` would still potentially lead to incorrect results. I think the warning could be more useful in that sense, although that isn't part of NumPy and one could very much argue this should be known by users.

Sure, in numpy, but in numpy you can mutate the underlying object, so later-on there is no way to know whether the individual elements have been assigned to or not. While in dask you are close to 100% sure that if an Array is directly derived from empty, it should not be read from.

It's a similar reasoning than the broadcast trick here; you can't generally use it in numpy, but here as array are immutable you can.

**Comment by jakirkham** [comment]
FWIW sometimes people do expect to do in-place operations with Dask ( https://github.com/dask/distributed/issues/1978 ).

Though agree `empty` would be a weird thing for most people to use in Dask. Maybe it would come about for compatibility reasons (trying to support various array-like libraries with the same code)?

Anyways maybe this is a little off-topic (albeit interesting to think about).

Happy to merge (if we are happy with this?) 🙂

**Comment by pentschev** [comment]
> Though agree empty would be a weird thing for most people to use in Dask. Maybe it would come about for compatibility reasons (trying to support various array-like libraries with the same code)?

Yeah, I don't know what would be the appropriate solution here, but given the previous comments I wonder if maybe any usage of `empty` in Dask should raise a warning. Also agree it's a bit off-topic, so happy to keep this as is.

**Comment by Carreau** [comment]
> Happy to merge (if we are happy with this?)

I am, just wasn't sure how to add tests that this actually decrease worker communications if you think it is necessary. Not sure if you also want some mention of this in the docs, and if so where ?

**Comment by Carreau** [comment]
Opened #6511 to track the  `empty()` issue.

**Comment by jakirkham** [comment]
> > Happy to merge (if we are happy with this?)
> 
> I am, just wasn't sure how to add tests that this actually decrease worker communications if you think it is necessary. Not sure if you also want some mention of this in the docs, and if so where ?

Probably would do that in Distributed. Could do this as a follow-up exercise. Though maybe starting with an issue is good if we are unsure how to test it?

**Comment by jsignell** [comment]
Yeah I don't think we need to block this on tests.

**Comment by jakirkham** [comment]
Thanks all! 😄 

Went ahead and merged. Hope that is ok 🙂 Let's follow up on anything else in subsequent PRs 😉

**Comment by Carreau** [comment]
Thanks !

**Pull request merged**